### ISSUE TITLE
Objects as records

### ISSUE THREAD

**Issue by frostburn** [xenharmonic-devs/sonic-weave#197]
JS object model so that `const {foo, bar} = baz` is legal syntax.

Data type:
```ts
type SonicWeaveRecord = Record<string, SonicWeaveValue>;
```

Literal syntax:
```js
{third: 5/4, fif: 3/2, "period or equave": 2}
```

Shorthand syntax:
```js
const third = 5/4
const fif = 3/2
const myRecord = {third, fif}
```

Empty syntax (precedence over empty block statement where legal):
```js
let emptyRecord = {}
```

Access syntax:
```js
myRecord["third"]
```

Spread:
```js
const clone = {...myRecord}
```

Destructuring declaration:
```js
let {third, fif} = myRecord
```

Destructuring assignment:
```js
{fif} = myRecord
```

Destructuring re-assignment:
```js
{third} *= myRecord
```

Checking for keys:
```js
"fif" in myRecord
```

Checking for values:
```js
3/2 of myRecord
```

Iterating over keys:
```js
for (const key in myRecord) {}
```

Iterating over values:
```js
for (const value of myRecord) {}
```

Iterating over entries:
```js
for (const [key, value] inof myRecord) {}
```

Builtins:
```
keys(myRecord) // ["third", "fif"] (unordered)
values(myRecord) // [3/2, 5/4] (unordered)
```

Keyword arguments (with #172):
```js
myRiff(1, 2, someArg=3, ...myRecord)
/* myRiff(1, 2, someArg=3, third=5/4, fif=3/2) */
```

Keyword rest parameters:
```js
riff myRiff(positionalParam, ...restParam, keywordOnlyParam, keywordOnlyParamWithDefault=3/2, ...restKeywordsParam) {}
```

Keyword test parameters (minimal):
```js
riff myRiff(..., ...restKeywordsParam) {}
```

**Comment by arseniiv** [comment]
This will be complicated, how much is it really needed? 🤔 

Also some things are quite cryptic, say `{third} *= myRecord`, what does happen here?

**Comment by frostburn** [comment]
> This will be complicated, how much is it really needed? 🤔

Only literal and access syntax are needed to get something useful for e.g. realizing step patterns.

> Also some things are quite cryptic, say `{third} *= myRecord`, what does happen here?

Let's compare:
```js
const myComma = 81/80
let myThird = 32/27
myThird *= myComma
print(myThird) // 6/5
```

```js
const myRecord = {myThird: 81/80}
let myThird = 32/27
{myThird} *= myRecord
print(myThird) // 6/5
```
Still esoteric, but doesn't collide with anything in the grammar so there's no technical reason this couldn't be implemented.

**Comment by arseniiv** [comment]
I understand better now. 👍 And what if we want to destructure a field named `a` into a variable named `b`? Would it be `{a: b} = record`? Looks cryptic but I don’t know how to do better.

**Comment by frostburn** [comment]
> I understand better now. 👍 And what if we want to destructure a field named `a` into a variable named `b` (and also other fields in other variables, so as to use destructuring syntax)? Would it be `{a: b} = record`? Looks cryptic but I don’t know how to do better.

Good catch! Forgot to list that one. Yeah, we're doing JS objects and their quirks here with some extra :hot_pepper: . But not pathological quirks like `'toString' in everyObject` evaluating to `true`.

**Comment by frostburn** [comment]
Record indexing by arrays of strings.
```js
const A = ["s", "M"]
const B = {s: 9/8, L: 10/9, M: 5/4}
B[A]  // [9/8, 5/4]
```
